File: Week 02/id_676/144.二叉树的前序遍历.java
import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

import javax.swing.tree.TreeNode;

/*
 * @lc app=leetcode.cn id=144 lang=java
 *
 * [144] 二叉树的前序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

 /*
 *思路：1.使用递归的方法实现，时间复杂度为O(n)
 * 2.使用栈的方法实现，时间复杂度为O(n)
 */

 //使用递归来实现
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        helper(root, res);
        return res;
    }

    public void helper(TreeNode root,List<Integer> res) {
        if (root != null) {
            res.add(root.val);
            if  (root.left != null ) {
                helper(root.left, res);
            }
            if (root.right != null) {
                helper(root.right, res);
            }
        }
    }
}
// @lc code=end

//使用栈来实现
class SolutionOne {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        Stack<TreeNode> stack = new Stack<>();
        TreeNode cur = root;
        while (!stack.isEmpty() || cur != null) {
            while (cur != null) {
                res.add(cur.val);
                stack.push(cur);
                cur = cur.left;
            }
            cur = stack.pop();
            cur = cur.right;
        }
        return res;
    }
}

//递归实现中序遍历
class SolutionTwo {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<>() ;
        helper(root, res);
        return res;
    }

    public void helper(TreeNode root,List<Integer> res) {
        if (root != null) {
            if (root.left != null) {
                helper(root.left, res);
            }
            res.add(root.val);
            if (root.right != null) {
                helper(root.right, res);
            }
        }
    }
}

//递归实现后序遍历
class SolutionThree{
    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        helper(root, res);
        return res;
    }

    public void helper(TreeNode root, List<Integer> res) {
        if (root != null) {
            if (root.left != null) {
                helper(root.left, res);
            }

            if (root.right != null) {
                helper(root.right, res);
            }
            res.add(root.val);
        }
    }
}

File: Week 02/id_521/144.二叉树的前序遍历.js
/*
 * @lc app=leetcode.cn id=144 lang=javascript
 *
 * [144] 二叉树的前序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var preorderTraversal = function(root) {
    let arr = []
    function helper(root, arr) {
        if(root !== null){
            arr.push(root.val); // 处理本层逻辑
            if(root.left !== null) {
                helper(root.left, arr) // 进入下一层
            }
            if(root.right !== null) { // 进入下一层， 这里其实有点分治的意思，左右分别处理
                helper(root.right, arr)
            }
        } 
    }
    helper(root, arr)
    return arr
};
// @lc code=end

